tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/GraphQL/HotChocolate/ASM/HotChocolateSecurity.cs

```diff
@@ -0,0 +1,175 @@
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using Datadog.Trace.AppSec;
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.GraphQL.HotChocolate.ASM.AST;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.GraphQL.HotChocolate.ASM;
+
+internal abstract class HotChocolateSecurity
+{
+    public static void ScanQuery(IQueryRequest request)
+    {
+        if (Tracer.Instance.ActiveScope is null)
+        {
+            return;
+        }
+
+        // Don't run if ASM isn't enabled or a WebSocket
+        if (!GraphQLSecurityCommon.IsEnabled())
+        {
+            return;
+        }
+
+        var doc = request.Query;
+        if (!doc.TryDuckCast<QueryDocument>(out var document))
+        {
+            return;
+        }
+
+        // Get root node of the document and do a depth search of resolvers
+        foreach (var node in document.Document.Definitions)
+        {
+            try
+            {
+                DepthSearchOperationNode(node, request.VariableValues);
+            }
+            catch
+            {
+                // Failed to search for resolvers with arguments
+            }
+        }
+    }
+
+    private static void DepthSearchOperationNode(object obj, IDictionary<string, object> variables)
+    {
+        if (!obj.TryDuckCast<SyntaxNode>(out var node))
+        {
+            return;
+        }
+
+        switch (node.Kind)
+        {
+            case SyntaxKindProxy.OperationDefinition when obj.TryDuckCast<SelectionsNode>(out var opNode):
+            {
+                foreach (var selectionSetSelection in opNode.SelectionSet.Selections)
+                {
+                    DepthSearchOperationNode(selectionSetSelection, variables);
+                }
+
+                break;
+            }
+
+            case SyntaxKindProxy.Field when obj.TryDuckCast<FieldNode>(out var fieldNode):
+            {
+                try
+                {
+                    if (fieldNode.Arguments.Any())
+                    {
+                        var resolverName = GetFieldNodeName(fieldNode, obj);
+                        var resolverArguments = new Dictionary<string, object>();
+
+                        foreach (var argument in fieldNode.Arguments)
+                        {
+                            var name = GetName(argument);
+                            var value = GetArgumentValue(argument, variables);
+                            resolverArguments.Add(name, value);
+                        }
+
+                        GraphQLSecurityCommon.RegisterResolverCall(Tracer.Instance.ActiveScope, resolverName, resolverArguments);
+                    }
+                }
+                catch
+                {
+                    // Failed to get argument value and register the resolver
```

Comment:
Same comment as above, logging the error generally makes sense unless it's something we expect very frequently